zaza/openstack/charm_tests/ovn/tests.py

```diff
@@ -896,6 +896,10 @@ def test_cluster_downscale(self):
         application = zaza.model.get_application("ovn-central")
 
         logging.info("Removing unit that hosts OVN follower server.")
+        # Run `update-status` hook. This updates the workload status message,
+        # helping us to correctly identify unit that does not host OVN leader
+        # servers.
+        zaza.run(application.run("hooks/update-status"))
```

Comment:
This does appear a bit brittle, I see one failure in the gate, and I had one failure locally after running the tests a couple of times.

One thing is that multiple charm units can have the "leader:" text in their status, the Northbound and Sothbound leader does not need to reside on the same unit.

Another thing is that juju might not be done executing when you return from the above call.

An alternative could be to collect this information with the cluster-status action you've just added?

Reply:
> One thing is that multiple charm units can have the "leader:" text in their status, the Northbound and Sothbound leader does not need to reside on the same unit.

I don't think that this is an issue, we just need a unit which hosts at least one leader. If the cluster downscales cleanly, it means that the graceful departure works even if we remove unit with leader server (it doesn't necessarily need to host both leaders).

> Another thing is that juju might not be done executing when you return from the above call.

Hmm, I assumed that [`Application.run()`](https://pythonlibjuju.readthedocs.io/en/latest/api/juju.application.html#juju.application.Application.run) waits for hook to finish given that it has also an option for timeout.

> An alternative could be to collect this information with the cluster-status action you've just added?

Definitely. This just seemed way more simple at the time.

>  I see one failure in the gate, and I had one failure locally after running the tests a couple of times.

Any chance you still have logs of the failure? I'd like to examine it and the OSCI seems to fully pass with these updated tests.

Reply:
> > Another thing is that juju might not be done executing when you return from the above call.
> 
> Hmm, I assumed that [`Application.run()`](https://pythonlibjuju.readthedocs.io/en/latest/api/juju.application.html#juju.application.Application.run) waits for hook to finish given that it has also an option for timeout.

The difference here is that after the hook tool has finished the global view of the Juju model status, on which your code depends, will update asynchronously. So for this to work, you most likely need to wait for something to happen+wait for things to be idle again before the status is accurate.

> > I see one failure in the gate, and I had one failure locally after running the tests a couple of times.
> 
> Any chance you still have logs of the failure? I'd like to examine it and the OSCI seems to fully pass with these updated tests.

I unfortunately do not have the logs available.

Reply:
"leader/follower detection" now uses `cluster-status` action